pkg/importer/dataStream_test.go

```diff
@@ -7,416 +7,404 @@ import (
+		defer func() {
+			tempTestServer.Close()
+		}()
+
+		if !wantErr {
+			Expect(err).NotTo(HaveOccurred())
+			for _, r := range ds.Readers {
+				fmt.Fprintf(GinkgoWriter, "INFO: Reader type: %d\n", r.rdrType)
+			}
+			Expect(len(ds.Readers)).To(Equal(numRdrs))
+		} else {
+			Expect(err).To(HaveOccurred())
+		}
+	},
+		table.Entry("successfully construct a xz reader", "tinyCore.iso.xz", testfiles[".xz"], 5, false),              // [http, multi-r, xz, multi-r]
+		table.Entry("successfully construct a gz reader", "tinyCore.iso.gz", testfiles[".gz"], 5, false),              // [http, multi-r, gz, multi-r]
+		table.Entry("successfully construct qcow2 reader", "", filepath.Join(imageDir, "cirros-qcow2.img"), 2, false), // [http, multi-r]
+		table.Entry("successfully construct .iso reader", "", filepath.Join(imageDir, "tinyCore.iso"), 3, false),      // [http, multi-r]
+		table.Entry("fail constructing reader for invalid file path", "", filepath.Join(imageDir, "tinyCorebad.iso"), 0, true),
+	)
+})
+
+var _ = Describe("Copy", func() {
+	var ts *httptest.Server
+
+	BeforeEach(func() {
+		By("[BeforeEach] Creating test server")
+		ts = createTestServer(imageDir)
+	})
 
-func replaceQEMUOperations(replacement image.QEMUOperations, f func()) {
-	orig := qemuOperations
-	if replacement != nil {
-		qemuOperations = replacement
-		defer func() { qemuOperations = orig }()
+	AfterEach(func() {
+		By("[AfterEach] closing test server")
+		ts.Close()
+	})
+
+	table.DescribeTable("Image, with import source should", func(dest, endpt string, qemuOperations image.QEMUOperations, wantErr bool) {
+		By("Configuring end point")
+		if !strings.HasPrefix(endpt, "http") {
+			endpt = ts.URL + "/" + endpt
+		}
+		By(fmt.Sprintf("end point is %s", endpt))
+		defer os.RemoveAll(dest)
+		By("Replacing QEMU Operations")
+		replaceQEMUOperations(qemuOperations, func() {
+			By("Copying image")
+			err := CopyImage(dest, endpt, "", "")
+			if !wantErr {
+				Expect(err).NotTo(HaveOccurred())
+			} else {
+				Expect(err).To(HaveOccurred())
+			}
+		})
+	},
+		table.Entry("successfully copy local image", filepath.Join(os.TempDir(), "cdi-testcopy"), "tinyCore.iso", NewQEMUAllErrors(), false),
+		table.Entry("expect failure trying to copy non-existing local image", filepath.Join(os.TempDir(), "cdi-testcopy"), "tinyCoreBad.iso", NewQEMUAllErrors(), true),
+		table.Entry("successfully copy streaming image", "cirros-qcow2.raw", "cirros-qcow2.img", NewFakeQEMUOperations(errors.New("should not be called"), nil, nil), false),
+		table.Entry("streaming image qemu validation fails", "cirros-qcow2.raw", "cirros-qcow2.img", NewFakeQEMUOperations(nil, nil, errors.New("invalid image")), true),
+		table.Entry("streaming image qemu convert fails", "cirros-qcow2.raw", "cirros-qcow2.img", NewFakeQEMUOperations(nil, errors.New("exit 1"), nil), true),
+	)
+
+	table.DescribeTable("internal copy", func(r io.Reader, out string, qemu bool, qemuOperations image.QEMUOperations, wantErr bool) {
```

Comment:
Not sure where to put this comment but, imo, I think the `isHTTPQcow2` block in `*DataStream.copy` should  be moved to `func copy()`. The reason being that the lowest level copy func, which is `func copy()` should contain all of the logic to do the copy which includes qcow2 streaming.